### PR TITLE
[FSDP2] allow meta tensors during loading state dict and cpu offloading

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126305
* __->__ #126267

unit test: ``pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py``

with meta init and cpu offloading, we have meta tensors after`model.load_state_dict(assign=True, strict=False)`. This PR avoided calling `.cpu` on meta tensors otherwise it's a runtime error

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 1]
```diff
@@ -68,6 +68,42 @@ def _test_1d_state_dict_save_load(self, mlp_dim: int):
...
+        # to test loading with `strict=False`
+        state_dicts = []
+        for name, dtensor in model.named_parameters():
+            full_tensor = torch.randn(dtensor.size())
+            sharded_tensor = distribute_tensor(
+                full_tensor, dtensor.device_mesh, dtensor.placements
+            )
+            state_dicts.append({name: sharded_tensor})
+
+        # allow meta params during `reset_sharded_param`
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:95]: nit: maybe something like this?
```suggestion
        # check that we can load with some parameters still on meta device
```
`reset_sharded_param` is an implementation detail (and may be refactored without remembering to change this comment, e.g. if we were to remove the method altogether after padding local tensor)

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 2]
```diff
@@ -68,6 +68,42 @@ def _test_1d_state_dict_save_load(self, mlp_dim: int):
...
+        for sd in state_dicts:
+            model.load_state_dict(sd, assign=True, strict=False)
+
+        # lazy init without error
+        inp = torch.rand((mlp_dim, mlp_dim), device="cuda")
+        model(inp)
+
+        state_dicts = model.state_dict()
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:103]: nit: this one seems should just be one state dict (singular, not plural)
```suggestion
        state_dict = model.state_dict()
```

**Review by awgu — APPROVED** [review]
LGTM! I left some nits on the unit test.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 3]
```diff
@@ -245,7 +245,7 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
         self.padded_sharded_param_size = padded_sharded_param.size()
         if sharded_param.numel() > 0:
             padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
-        if self.offload_to_cpu:
+        if self.offload_to_cpu and not torch.empty(0).is_meta:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:248]:
```suggestion
        if self.offload_to_cpu and not padded_sharded_param.is_meta:
```

Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:248]: resolving in https://github.com/pytorch/pytorch/pull/126305
